### PR TITLE
torch._dynamo.reset() before and after running each test

### PR DESCRIPTION
**Pull request by zou3519**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #110107

Previously, running tests with PYTORCH_TEST_WITH_DYNAMO=1 ran some tests in eager-mode pytorch because the Dynamo cache filled up. We need to reset the Dynamo state before each test in order to actually run each test using Dynamo instead of non-deterministically falling back to eager-mode.

Test Plan:
- CI

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

### CODE REVIEW

File: torch/testing/_internal/common_utils.py [hunk 1]
```diff
@@ -2467,14 +2467,25 @@ def _run_with_retry(self, result=None, num_runs_left=0, report_only=True, num_re
...
+        def dynamo_wrapper(backend, function):
+            @functools.wraps(function)
+            def inner(*args, **kwargs):
+                # It's important to reset the Dynamo state so that previous
+                # invocations do not affect future ones.
+                torch._dynamo.reset()
+                try:
+                    return torch._dynamo.optimize(backend)(function)(*args, **kwargs)
+                finally:
+                    torch._dynamo.reset()
```
Comment (voznesenskym) [inline torch/testing/_internal/common_utils.py:2479]: Why after too?

**Review by voznesenskym — APPROVED** [review]

**Review by ezyang — APPROVED** [review]
You can ship this if it doesn't cause timeouts 🤔

File: torch/testing/_internal/common_utils.py (hunk 1 above)
Comment (zou3519) [reply torch/testing/_internal/common_utils.py:2479]: It's good to clean up after ourselves, but mechanically we don't really need it

**Comment by zou3519** [comment]
> You can ship this if it doesn't cause timeouts 🤔

I've solved the timeouts problem with more shards. The problem right now is skipping enough tests until the CI is green lol

File: torch/testing/_internal/common_utils.py (hunk 1 above)
Comment (voznesenskym) [reply torch/testing/_internal/common_utils.py:2479]: ok

**Review by malfet — APPROVED** [review]
Let's keep an eye on potentially increasing the queue time as we are using more runners now

**Comment by malfet** [comment]
> I've solved the timeouts problem with more shards. The problem right now is skipping enough tests until the CI is green lol

How much longer it takes with cache invalidation then without it?

**Comment by zou3519** [comment]
> > I've solved the timeouts problem with more shards. The problem right now is skipping enough tests until the CI is green lol
> 
> How much longer it takes with cache invalidation then without it?

The total time is now 2-3x of what it was previously. So far I have increased the shard count from 2 to 5

**Pull request closed without merging**